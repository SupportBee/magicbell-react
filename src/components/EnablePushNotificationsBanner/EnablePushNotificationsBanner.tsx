/** @jsx jsx */
import { css, jsx } from '@emotion/react';
import { clientSettings, useConfig } from '@magicbell/react-headless';
import axios from 'axios';
import get from 'lodash/get';
import { useLocalStorage } from 'react-use';
import { useTheme } from '../../context/MagicBellThemeContext';
import { toRGBA } from '../../lib/color';
import { openWindow } from '../../lib/window';
import Text from '../Text';
import CloseBannerButton from './CloseBannerButton';
import EnablePushNotificationsButton from './EnablePushNotificationsButton';

/**
 * Banner to ask the user to enable push notifications.
 *
 * @example
 * <EnablePushNotificationsBanner />
 */
export default function EnablePushNotificationsBanner() {
  const { getState } = clientSettings;
  const { apiKey, userEmail, userExternalId } = getState();
  const config = useConfig();
  const theme = useTheme();
  const { footer: footerTheme } = theme;
  const { notification: notificationTheme } = theme;
  const [wasRequested, setRequestedAt] = useLocalStorage<number | null>(
    `magicbell:${apiKey}:web-push-requested-at`,
    null,
  );
  const isWebPushEnabled = get(config.channels, 'webPush.enabled');

  const enablePushNotifications = () => {
    const baseUrl = get(config.channels, 'webPushNotifications.subscribeUrl');
    const url = axios.getUri({
      url: baseUrl,
      params: {
        user_email: userEmail,
        user_external_id: userExternalId,
        background_color: theme.header.backgroundColor,
        text_color: theme.header.textColor,
      },
    });

    setRequestedAt(Date.now());
    openWindow(url);
  };

  const closeBanner = () => {
    setRequestedAt(Date.now());
  };

  if (wasRequested || !isWebPushEnabled) return null;
  return (
    <div
      css={css`
        padding: 14px 24px !important;
        display: flex;
        text-align: left;
        align-items: center;
        background: ${toRGBA(notificationTheme.unseen.backgroundColor, 0.1)};

        color: ${notificationTheme.unseen.textColor} !important;
        font-family: ${footerTheme.fontFamily} !important;
        text-align: ${footerTheme.textAlign} !important;
        font-size: ${footerTheme.fontSize} !important;

        & > * {
          margin-left: 1rem;
        }
      `}
    >
      <p style={{ flex: 1 }}>
        <Text
          id="web-push-notifications.notice"
          defaultMessage="By enabling browser notifications, youâ€™ll stay up to date even better."
        />
      </p>
      <EnablePushNotificationsButton onClick={enablePushNotifications} />
      <CloseBannerButton onClick={closeBanner} />
    </div>
  );
}
