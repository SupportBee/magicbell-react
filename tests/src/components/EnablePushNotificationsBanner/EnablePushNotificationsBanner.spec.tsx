import { clientSettings, useConfig } from '@magicbell/react-headless';
import { act, fireEvent, render, RenderResult } from '@testing-library/react';
import React from 'react';
import EnablePushNotificationsBanner from '../../../../src/components/EnablePushNotificationsBanner/EnablePushNotificationsBanner';
import ConfigFactory, { sampleConfig } from '../../../factories/ConfigFactory';

describe('components', () => {
  describe('EnablePushNotificationsBanner', () => {
    let view: RenderResult;

    beforeEach(() => {
      clientSettings.setState({ apiKey: 'cf63e9f2fcb30bcd58eb' });
      useConfig.setState({
        ...sampleConfig,
        lastFetchedAt: Date.now(),
      });
      localStorage.clear();

      view = render(<EnablePushNotificationsBanner />);
    });

    describe('render', () => {
      describe('the user did not enable push notifications before', () => {
        it('renders a button to enable push notifications', () => {
          expect(view.container).toMatchSnapshot();
        });
      });

      describe('the user has enabled push notifications before', () => {
        beforeEach(() => {
          localStorage.setItem(
            'magicbell:cf63e9f2fcb30bcd58eb:web-push-requested-at',
            JSON.stringify(Date.now()),
          );
          view = render(<EnablePushNotificationsBanner />);
        });

        afterEach(() => {
          localStorage.clear();
        });

        it('does not render anything', () => {
          expect(view.container).toMatchSnapshot();
        });
      });

      describe('the web push channel is disabled', () => {
        beforeEach(() => {
          act(() => {
            useConfig.setState({
              ...ConfigFactory.build({ channels: { webPush: { enabled: false } } }),
              lastFetchedAt: Date.now(),
            });
          });

          view = render(<EnablePushNotificationsBanner />);
        });

        it('does not render anything', () => {
          expect(view.container).toMatchSnapshot();
        });
      });
    });

    describe('.enablePushNotifications', () => {
      it('opens a new window to create the push subscription', () => {
        window.open = jest.fn();
        const button = view.getAllByRole('button')[0];
        fireEvent.click(button);

        expect(window.open).toHaveBeenCalledTimes(1);
      });
    });

    describe('.closeBanner', () => {
      it('closes the banner', () => {
        const button = view.getAllByRole('button')[1];
        fireEvent.click(button);

        expect(view.container).toMatchSnapshot();
      });
    });
  });
});
