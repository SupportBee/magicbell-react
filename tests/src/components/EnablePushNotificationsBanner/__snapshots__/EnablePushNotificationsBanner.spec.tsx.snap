// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`components EnablePushNotificationsBanner .closeBanner closes the banner 1`] = `<div />`;

exports[`components EnablePushNotificationsBanner render the user did not enable push notifications before renders a button to enable push notifications 1`] = `
.emotion-0 {
  padding: 14px 24px!important;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  text-align: left;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  background: rgba(52, 152, 244, 0.1);
  color: #3A424D!important;
  font-family: -apple-system,BlinkMacSystemFont,"Segoe UI",Roboto,Oxygen-Sans,Ubuntu,Cantarell,"Helvetica Neue",sans-serif!important;
  text-align: left!important;
  font-size: 14px!important;
}

.emotion-0>* {
  margin-left: 1rem;
}

.emotion-1 {
  padding: 10px!important;
  border-radius: 4px!important;
  background: #3498F4!important;
  color: white!important;
}

.emotion-2 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
}

.emotion-2 svg {
  height: 24px;
}

<div>
  <div
    class="emotion-0"
  >
    <p
      style="flex: 1;"
    >
      By enabling browser notifications, youâ€™ll stay up to date even better.
    </p>
    <button
      class="emotion-1"
    >
      Enable Now
    </button>
    <button
      class="emotion-2"
    >
      <svg
        fill="none"
        stroke="currentColor"
        viewBox="0 0 24 24"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M6 18L18 6M6 6l12 12"
          stroke-linecap="round"
          stroke-linejoin="round"
          stroke-width="2"
        />
      </svg>
    </button>
  </div>
</div>
`;

exports[`components EnablePushNotificationsBanner render the user has enabled push notifications before does not render anything 1`] = `<div />`;

exports[`components EnablePushNotificationsBanner render the web push channel is disabled does not render anything 1`] = `<div />`;
